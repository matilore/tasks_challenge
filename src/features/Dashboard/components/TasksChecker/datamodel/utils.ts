interface TasksGroup {
  name: string
  tasks: Task[]
}

interface Task {
  description: string
  value: number
  checked: boolean
}

export const getSum = (tasks: TasksGroup[]): number => {
  const tasksToParse = [...tasks]
  return tasksToParse.reduce((acc, el) => {
    if (el.tasks) {
      el.tasks.forEach(({ value }) => (acc += value))
    }
    return acc
  }, 0)
}

export const parseResponse = (tasks: TasksGroup[]) => {
  const sum = getSum(tasks)
  let normalizedSum = 0
  const parsedData = tasks.map(({ name, tasks }, index) => {
    const parsedTasks = tasks.map(({ value, ...task }) => {
      const normalizedValue = Math.round((value * 100) / sum)
      normalizedSum += task.checked ? normalizedValue : 0
      return {
        ...task,
        value: normalizedValue
      }
    })
    return { id: index + 1, name, tasks: parsedTasks }
  })
  return { data: parsedData, normalizedSum }
}
