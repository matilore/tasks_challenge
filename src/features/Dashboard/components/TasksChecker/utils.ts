import produce from 'immer'
import { TasksGroup } from './contracts'

export const getNextTasksState = (
  tasksList: TasksGroup[],
  groupIndex: number,
  index: number
) => {
  return produce(tasksList, draftState => {
    draftState[groupIndex].tasks[index].checked =
      !draftState[groupIndex].tasks[index].checked
  })
}
