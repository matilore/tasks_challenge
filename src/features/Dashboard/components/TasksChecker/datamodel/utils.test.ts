import { getSum, parseResponse } from './utils'

describe('getSum', () => {
  it('returns the correct progress of task values', () => {
    const tasks = [
      {
        name: 'Group 1',
        tasks: [{ description: 'lorem ipsum', value: 10, checked: true }]
      },
      {
        name: 'Group 2',
        tasks: [{ description: 'lorem ipsum', value: 30, checked: true }]
      }
    ]
    const result = getSum(tasks)
    expect(result).toEqual(40)
  })
})

describe('parseResponse', () => {
  it('correctly parses the response data', () => {
    const tasks = [
      {
        name: 'Group 1',
        tasks: [
          { description: 'Task 1', value: 10, checked: true },
          { description: 'Task 2', value: 20, checked: false },
          { description: 'Task 3', value: 30, checked: true }
        ]
      },
      {
        name: 'Group 2',
        tasks: [
          { description: 'Task 4', value: 15, checked: false },
          { description: 'Task 5', value: 25, checked: true }
        ]
      }
    ]
    const result = parseResponse(tasks)
    expect(result).toMatchSnapshot()
  })
})
