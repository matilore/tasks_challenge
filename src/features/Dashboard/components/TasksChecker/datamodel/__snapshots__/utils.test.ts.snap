// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseResponse correctly parses the response data 1`] = `
{
  "data": [
    {
      "id": 1,
      "name": "Group 1",
      "tasks": [
        {
          "checked": true,
          "description": "Task 1",
          "value": 10,
        },
        {
          "checked": false,
          "description": "Task 2",
          "value": 20,
        },
        {
          "checked": true,
          "description": "Task 3",
          "value": 30,
        },
      ],
    },
    {
      "id": 2,
      "name": "Group 2",
      "tasks": [
        {
          "checked": false,
          "description": "Task 4",
          "value": 15,
        },
        {
          "checked": true,
          "description": "Task 5",
          "value": 25,
        },
      ],
    },
  ],
  "normalizedSum": 65,
}
`;
